diff --git a/css/styles.css b/css/styles.css
index acd1b6c..db74af0 100644
--- a/css/styles.css
+++ b/css/styles.css
@@ -41,6 +41,17 @@ body {
    
 }
 
+.keyword {
+   color: #ffffff;
+   background-color: #0088cc;
+   padding: 8px 10px;
+   margin: 2px;
+   -webkit-border-radius: 5px;
+   -moz-border-radius: 5px;
+   border-radius: 5px;
+   line-height: 36px;
+}
+
 ul, li, dt, dl, dd, form, img, input {
    margin: 0;
    padding: 0;
diff --git a/front/knowbaseitem.form.php b/front/knowbaseitem.form.php
index a3d750f..489d68e 100644
--- a/front/knowbaseitem.form.php
+++ b/front/knowbaseitem.form.php
@@ -51,21 +51,32 @@ $kb = new KnowbaseItem();
 if (isset($_POST["add"])) {
    // ajoute un item dans la base de connaisssances
    $kb->check(-1,'w',$_POST);
-
-   $newID = $kb->add($_POST);
-   Event::log($newID, "knowbaseitem", 5, "tools",
-              sprintf(__('%1$s adds the item %2$s'), $_SESSION["glpiname"], $newID));
-   Html::redirect($CFG_GLPI["root_doc"]."/front/knowbaseitem.php");
-
+   
+   if (empty($_REQUEST['keywords'])) {
+      Session::addMessageAfterRedirect(__('Keywords is a mandatory field'), false, ERROR);
+      Html::back();
+   } else {
+      
+      $newID = $kb->add($_POST);
+      Event::log($newID, "knowbaseitem", 5, "tools",
+                 sprintf(__('%1$s adds the item %2$s'), $_SESSION["glpiname"], $newID));
+      Html::redirect($CFG_GLPI["root_doc"]."/front/knowbaseitem.php");
+   }
 } else if (isset($_POST["update"])) {
    // actualiser  un item dans la base de connaissances
    $kb->check($_POST["id"],'w');
-
-   $kb->update($_POST);
-   Event::log($_POST["id"], "knowbaseitem", 5, "tools",
-              //TRANS: %s is the user login
-              sprintf(__('%s updates an item'), $_SESSION["glpiname"]));
-   Html::redirect($CFG_GLPI["root_doc"]."/front/knowbaseitem.form.php?id=".$_POST['id']);
+   
+   if (empty($_REQUEST['keywords']) && isset($_REQUEST['modify'])) {
+      Session::addMessageAfterRedirect(__('Keywords is a mandatory field'), false, ERROR);
+      Html::back();
+   } else {
+      
+      $kb->update($_POST);
+      Event::log($_POST["id"], "knowbaseitem", 5, "tools",
+                 //TRANS: %s is the user login
+                 sprintf(__('%s updates an item'), $_SESSION["glpiname"]));
+      Html::redirect($CFG_GLPI["root_doc"]."/front/knowbaseitem.form.php?id=".$_POST['id']);
+   }
 
 } else if (isset($_POST["delete"])) {
    // effacer un item dans la base de connaissances
diff --git a/inc/knowbase.class.php b/inc/knowbase.class.php
index 1dcd89f..3ce8bd2 100644
--- a/inc/knowbase.class.php
+++ b/inc/knowbase.class.php
@@ -121,8 +121,8 @@ class Knowbase extends CommonGLPI {
       }
       $ki = new KnowbaseItem();
       $ki->searchForm($_POST);
-
-      if (!isset($_POST['contains']) || empty($_POST['contains'])) {
+      
+      if (!isset($_POST['knowbaseitemcategories_id'])) {
          echo "<div><table class='center-h' width='950px'><tr><td class='center top'>";
          KnowbaseItem::showRecentPopular("recent");
          echo "</td><td class='center top'>";
@@ -131,8 +131,10 @@ class Knowbase extends CommonGLPI {
          KnowbaseItem::showRecentPopular("popular");
          echo "</td></tr>";
          echo "</table></div>";
-      } else {
+      } else if (isset($_POST['contains']) && !isset($_POST['contains_keywords'])) {
          KnowbaseItem::showList($_POST, 'search');
+      } else if (isset($_POST['contains_keywords'])) {
+         KnowbaseItem::showList($_POST, 'search_keywords');
       }
    }
 
diff --git a/inc/knowbaseitem.class.php b/inc/knowbaseitem.class.php
index ca55fbc..ee31d94 100644
--- a/inc/knowbaseitem.class.php
+++ b/inc/knowbaseitem.class.php
@@ -566,7 +566,8 @@ class KnowbaseItem extends CommonDBTM {
          echo "<div id='contenukb'>";
          Html::initEditorSystem('answer');
 
-         echo "<form method='post' id='form_kb' name='form_kb' action=\"".$this->getFormUrl()."\">";
+         echo "<form method='post' id='form_kb' name='form_kb' action=\"".$this->getFormUrl()."\" 
+               onsubmit='return check();'>";
 
          if (!empty($ID)) {
             echo "<input type='hidden' name='id' value='$ID'>\n";
@@ -577,6 +578,30 @@ class KnowbaseItem extends CommonDBTM {
          echo "<div class='center'>".__('Select a category for this item')."&nbsp;";
          KnowbaseItemCategory::dropdown(array('value' => $this->fields["knowbaseitemcategories_id"]));
          echo "</div></fieldset>";
+         
+         echo "<fieldset>";
+         echo "<legend>".__('Keywords')."</legend>";
+         echo "<div class='center'>";
+         echo __('Keywords separated by commas')."&nbsp;";
+         echo "<span style='color:red'>*</span>";         
+         echo "<textarea cols='80' rows='2' name='keywords' required>";
+         self::showKeywords();
+         echo "</textarea>";
+         echo "</div></fieldset>";
+         
+         //TODO : Finir de rendre le champ 'keyword' obligatoire #SIMPLE
+         echo "<script>
+               // For old browser :
+               function check() {
+                  var actual_value = document.getElementsByName('keywords')[0].value;
+                  if (actual_value == '') {
+                     alert('".__('Keywords is a mandatory field')."');
+                     return false;
+                  } else {
+                     return true;
+                  }
+               }
+               </script>";
 
          echo "<fieldset>";
          echo "<legend>".__('Subject')."</legend>";
@@ -763,6 +788,10 @@ class KnowbaseItem extends CommonDBTM {
       $DB->query($query);
    }
 
+   
+   function showKeywords() {
+      echo $this->fields["keywords"];
+   }
 
    /**
     * Print out (html) show item : question and answer
@@ -815,6 +844,16 @@ class KnowbaseItem extends CommonDBTM {
       echo "</div>";
       echo "</td></tr>";
 
+      echo "<tr class='tab_bg_3'><td class='left' colspan='4'><h2>".__('Keywords')."</h2>";
+      echo "<br>";
+      if ($this->fields["keywords"] != '') { // (Old KB_item no have keywords)
+         foreach (explode(",", $this->fields["keywords"]) as $keyword) {
+            echo "<span class='keyword'>".trim($keyword)."</span>";
+         }
+      }
+      echo "<br>";
+      echo "</th></tr>";
+      
       echo "<tr><th class='tdkb'  colspan='2'>";
       if ($this->fields["users_id"]) {
          // Integer because true may be 2 and getUserName return array
@@ -887,10 +926,25 @@ class KnowbaseItem extends CommonDBTM {
 
       echo "<form method='get' action='".$this->getSearchURL()."'>";
       echo "<table class='tab_cadre_fixe'>";
-      echo "<tr class='tab_bg_2'><td class='right' width='50%'>";
-      echo "<input type='text' size='50' name='contains' value=\"".
-             stripslashes(Html::cleanInputText($params["contains"]))."\"></td>";
+      echo "<tr class='tab_bg_2'><td class='right' width='60%'>";
+      echo __("General search"). " : ";
+
+      echo "<input type='text' size='50' name='contains' value=\"";
+      if (!isset($_REQUEST["other_form"])) {
+         echo stripslashes(Html::cleanInputText($params["contains"]));
+      }
+      echo "\"></td>";
       echo "<td class='left'>";
+      
+      echo __('Category');
+      if (isset($_REQUEST['knowbaseitemcategories_id'])) {
+         $knowbaseitemcategories_id = $_REQUEST['knowbaseitemcategories_id'];
+      } else {
+         $knowbaseitemcategories_id = '';
+      }
+      KnowbaseItemCategory::dropdown(array('value' => $knowbaseitemcategories_id));
+      echo "&nbsp;";
+      
       echo "<input type='submit' value=\""._sx('button','Search')."\" class='submit'></td></tr>";
       echo "</table>";
       if (isset($options['item_itemtype'])
@@ -901,6 +955,35 @@ class KnowbaseItem extends CommonDBTM {
       Html::closeForm();
 
       echo "</div>";
+      
+      echo "<div>";
+      
+      echo "<form method='get' action='".$this->getSearchURL()."'>";
+      echo "<table class='tab_cadre_fixe'>";
+      echo "<tr class='tab_bg_2'><td class='right' width='60%'>";
+      echo __("Keywords"). " : ";
+      echo "<input type='text' size='50' name='contains_keywords' value=\"";
+      if (isset($_REQUEST["other_form"])) {
+         echo stripslashes(Html::cleanInputText($params["contains_keywords"]));
+      }
+      echo "\"></td>";
+      echo "<td class='left'>";
+      
+      echo __('Category');
+      KnowbaseItemCategory::dropdown(array('value' => $knowbaseitemcategories_id));
+      echo "&nbsp;";
+      
+      echo "<input type='hidden' name='other_form' value='yes'>";
+      echo "<input type='submit' value=\""._sx('button','Search')."\" class='submit'></td></tr>";
+      echo "</table>";
+      if (isset($options['item_itemtype'])
+            && isset($options['item_items_id'])) {
+               echo "<input type='hidden' name='item_itemtype' value='".$options['item_itemtype']."'>";
+               echo "<input type='hidden' name='item_items_id' value='".$options['item_items_id']."'>";
+            }
+            Html::closeForm();
+      
+      echo "</div>";
    }
 
 
@@ -1009,6 +1092,19 @@ class KnowbaseItem extends CommonDBTM {
    **/
    static function getListRequest(array $params, $type='search') {
       global $DB;
+      
+      switch ($type) {
+         case "search_keywords" :
+            if (strlen($params["contains_keywords"]) == 0) {
+               $params["contains_keywords"] = "*";
+            }
+            break;
+         case "search" :
+            if (strlen($params["contains"]) == 0) {
+               $params["contains"] = "*";
+            }
+            break;
+      }
 
       // Lists kb Items
       $where = "";
@@ -1070,6 +1166,47 @@ class KnowbaseItem extends CommonDBTM {
                               AND `glpi_knowbaseitems_users`.`users_id` IS NULL)";
             break;
 
+         case 'search_keywords' :
+                        if (strlen($params["contains_keywords"]) > 0) {
+              $search  = Toolbox::unclean_cross_side_scripting_deep($params["contains_keywords"]);
+
+               $score   = " ,MATCH(`glpi_knowbaseitems`.`keywords`)
+                           AGAINST('$search' IN BOOLEAN MODE) AS SCORE ";
+
+               if (isset($_REQUEST['knowbaseitemcategories_id']) && !empty($_REQUEST['knowbaseitemcategories_id'])) {
+                  $where .= " AND `glpi_knowbaseitems`.`knowbaseitemcategories_id` = ".$_REQUEST['knowbaseitemcategories_id']." ";
+               }
+               
+               $where_1 = $where." AND MATCH(`glpi_knowbaseitems`.`keywords`) AGAINST('$search' IN BOOLEAN MODE) ";
+               $order   = "ORDER BY `SCORE` DESC";
+
+               // preliminar query to allow alternate search if no result with fulltext
+               $query_1   = "SELECT COUNT(`glpi_knowbaseitems`.`id`)
+                             FROM `glpi_knowbaseitems`
+                             $join
+                             WHERE $where_1";
+               $result_1  = $DB->query($query_1);
+               $numrows_1 = $DB->result($result_1,0,0);
+
+               if ($numrows_1 <= 0) {// not result this fulltext try with alternate search
+                  $search1 = array(/* 1 */   '/\\\"/',
+                                   /* 2 */   "/\+/",
+                                   /* 3 */   "/\*/",
+                                   /* 4 */   "/~/",
+                                   /* 5 */   "/</",
+                                   /* 6 */   "/>/",
+                                   /* 7 */   "/\(/",
+                                   /* 8 */   "/\)/",
+                                   /* 9 */   "/\-/");
+                  $contains = preg_replace($search1,"", $params["contains_keywords"]);
+                  $addwhere = '';
+                  $where   .= " AND (`glpi_knowbaseitems`.`keywords` ".Search::makeTextSearch($contains)." )";
+               } else {
+                  $where = $where_1;
+               }
+            }
+            break;
+
          case 'search' :
             if (strlen($params["contains"]) > 0) {
                $search  = Toolbox::unclean_cross_side_scripting_deep($params["contains"]);
@@ -1077,7 +1194,11 @@ class KnowbaseItem extends CommonDBTM {
                $score   = " ,MATCH(`glpi_knowbaseitems`.`name`, `glpi_knowbaseitems`.`answer`)
                            AGAINST('$search' IN BOOLEAN MODE) AS SCORE ";
 
-               $where_1 = $where." AND MATCH(`glpi_knowbaseitems`.`name`, `glpi_knowbaseitems`.`answer`)
+               if (isset($_REQUEST['knowbaseitemcategories_id']) && !empty($_REQUEST['knowbaseitemcategories_id'])) {
+                  $where .= " AND `glpi_knowbaseitems`.`knowbaseitemcategories_id` = ".$_REQUEST['knowbaseitemcategories_id']." ";
+               }
+               
+               $where_1 = $where." AND MATCH(`glpi_knowbaseitems`.`name`, `glpi_knowbaseitems`.`answer`, `glpi_knowbaseitems`.`keywords`)
                           AGAINST('$search' IN BOOLEAN MODE) ";
 
                $order   = "ORDER BY `SCORE` DESC";
@@ -1245,6 +1366,8 @@ class KnowbaseItem extends CommonDBTM {
                echo Search::showHeaderItem($output_type, __('Writer'), $header_num);
             }
             echo Search::showHeaderItem($output_type, __('Category'), $header_num);
+            
+            echo Search::showHeaderItem($output_type, __('Keywords'), $header_num);
 
             if (isset($options['item_itemtype'])
                 && isset($options['item_items_id'])
@@ -1309,6 +1432,9 @@ class KnowbaseItem extends CommonDBTM {
                   $categ   = "<a href='$cathref'>".$categ.'</a>';
                }
                echo Search::showItem($output_type, $categ, $item_num, $row_num);
+               
+               $categ = $data["keywords"];
+               echo Search::showItem($output_type, $categ, $item_num, $row_num);
 
 
                if (isset($options['item_itemtype'])
@@ -1492,6 +1618,12 @@ class KnowbaseItem extends CommonDBTM {
       $tab[86]['field']         = 'is_recursive';
       $tab[86]['name']          = __('Child entities');
       $tab[86]['datatype']      = 'bool';
+      
+      //TODO : A garder ?
+      $tab[100]['table']          = 'knowbaseitems';
+      $tab[100]['field']          = 'keywords';
+      $tab[100]['name']           = __('Keywords');
+      $tab[100]['datatype']       = 'dropdown'; //TODO
 
       return $tab;
    }
diff --git a/install/mysql/glpi-0.84.4-empty.sql b/install/mysql/glpi-0.84.4-empty.sql
index 6fab97b..710119c 100644
--- a/install/mysql/glpi-0.84.4-empty.sql
+++ b/install/mysql/glpi-0.84.4-empty.sql
@@ -2427,6 +2427,7 @@ CREATE TABLE `glpi_knowbaseitems` (
   `view` int(11) NOT NULL DEFAULT '0',
   `date` datetime DEFAULT NULL,
   `date_mod` datetime DEFAULT NULL,
+  `keywords` text COLLATE utf8_unicode_ci,
   PRIMARY KEY (`id`),
   KEY `users_id` (`users_id`),
   KEY `knowbaseitemcategories_id` (`knowbaseitemcategories_id`),
diff --git a/install/update_0831_084.php b/install/update_0831_084.php
index 49c6845..855d2f9 100644
--- a/install/update_0831_084.php
+++ b/install/update_0831_084.php
@@ -83,6 +83,8 @@ function update0831to084() {
    $migration->addField('glpi_mailcollectors', 'use_kerberos', 'bool', array('value' => 0));
    $migration->addField("glpi_mailcollectors", 'errors', "integer");
    $migration->addField("glpi_mailcollectors", 'use_mail_date', "bool", array('value' => 0));
+   
+   $migration->addField('glpi_knowbaseitems', 'keywords', 'text');
 
    // Password security
    $migration->addField('glpi_configs', 'use_password_security', 'bool');
diff --git a/locales/en_GB.po b/locales/en_GB.po
index 1c40a37..cd027d3 100644
--- a/locales/en_GB.po
+++ b/locales/en_GB.po
@@ -19,6 +19,18 @@ msgstr ""
 "Language: en_GB\n"
 "Plural-Forms: nplurals=2; plural=(n != 1);\n"
 
+#: inc/knowbaseitem.class.php:582
+msgid "Keywords"
+msgstr "Keywords"
+
+#: inc/knowbaseitem.class.php:584
+msgid "Keywords separated by commas"
+msgstr "Keywords separated by commas"
+
+#: front/knowbaseitem.form.php:62
+msgid "Keywords is a mandatory field"
+msgstr "Keywords is a mandatory field"
+
 #: index.php:69
 msgid "GLPI - Authentication"
 msgstr "GLPI - Authentication"
@@ -6648,6 +6660,9 @@ msgstr "Address"
 msgid "Fax"
 msgstr "Fax"
 
+msgid "General search"
+msgstr "General search"
+
 #: inc/contact.class.php:188 inc/contact.class.php:349
 #: inc/entity.class.php:570 inc/entity.class.php:991
 #: inc/supplier.class.php:161 inc/supplier.class.php:280
diff --git a/locales/fr_FR.po b/locales/fr_FR.po
index 8d2835b..8468bd7 100644
--- a/locales/fr_FR.po
+++ b/locales/fr_FR.po
@@ -27,6 +27,18 @@ msgstr ""
 "Language: fr_FR\n"
 "Plural-Forms: nplurals=2; plural=(n > 1);\n"
 
+#: inc/knowbaseitem.class.php:582
+msgid "Keywords"
+msgstr "Mots-clés"
+
+#: inc/knowbaseitem.class.php:584
+msgid "Keywords separated by commas"
+msgstr "Mots-clés séparés par des virgules"
+
+#: front/knowbaseitem.form.php:62
+msgid "Keywords is a mandatory field"
+msgstr "Mots-clés est un champ obligatoire"
+
 #: index.php:69
 msgid "GLPI - Authentication"
 msgstr "GLPI - Authentification"
@@ -6656,6 +6668,9 @@ msgstr "Adresse"
 msgid "Fax"
 msgstr "Fax"
 
+msgid "General search"
+msgstr "Recherche générale"
+
 #: inc/contact.class.php:188 inc/contact.class.php:349
 #: inc/entity.class.php:570 inc/entity.class.php:991
 #: inc/supplier.class.php:161 inc/supplier.class.php:280
diff --git a/locales/glpi.pot b/locales/glpi.pot
index 7009103..312cce5 100644
--- a/locales/glpi.pot
+++ b/locales/glpi.pot
@@ -18,6 +18,18 @@ msgstr ""
 "Content-Transfer-Encoding: 8bit\n"
 "Plural-Forms: nplurals=INTEGER; plural=EXPRESSION;\n"
 
+#: inc/knowbaseitem.class.php:582
+msgid "Keywords"
+msgstr ""
+
+#: inc/knowbaseitem.class.php:584
+msgid "Keywords separated by commas"
+msgstr ""
+
+#: front/knowbaseitem.form.php:62
+msgid "Keywords is a mandatory field"
+msgstr ""
+
 #: index.php:69
 msgid "GLPI - Authentication"
 msgstr ""